libs/hwes/eventstoredb/projections/custom/custom.go

```diff
@@ -0,0 +1,135 @@
+}
+
+func (p *CustomProjection) RegisterEventListener(eventType string, eventHandler eventHandler) *CustomProjection {
+	if _, found := p.eventHandlers[eventType]; found {
+		zlog.Error().
+			Str("eventType", eventType).
+			Msg("override existing event handler")
+	}
+
+	p.eventHandlers[eventType] = eventHandler
+	return p
+}
+
+func (p *CustomProjection) handleEvent(ctx context.Context, event hwes.Event) (error, esdb.Nack_Action) {
+	eventHandler, found := p.eventHandlers[event.EventType]
+	if !found {
+		return fmt.Errorf("event type '%s' is invalid", event.EventType), esdb.Nack_Unknown
+	}
+	return eventHandler(ctx, event)
+}
+
+func (p *CustomProjection) Subscribe(ctx context.Context) error {
+	log := zlog.Ctx(ctx)
+
+	// Create subscription on EventStoreDB
+	persistentAllSubscriptionOptions := esdb.PersistentAllSubscriptionOptions{}
+	err := p.es.CreatePersistentSubscriptionAll(ctx, subscriptionGroupName, persistentAllSubscriptionOptions)
+	if err != nil {
+		// Ignore subscription already exists error
+		if persistentSubscriptionError, ok := err.(*esdb.PersistentSubscriptionError); !ok || ok && (persistentSubscriptionError.Code != 6) {
+			zlog.Ctx(ctx).Warn().Err(err).Send()
+			return err
+		}
+	}
+
+	// Subscribe to server side subscription
+	stream, err := p.es.ConnectToPersistentSubscriptionToAll(ctx, subscriptionGroupName, esdb.ConnectToPersistentSubscriptionOptions{})
+	if err != nil {
+		return err
+	}
+	defer stream.Close()
+
+	log.Info().Msg("Start custom projection")
+	err = p.process(ctx, stream)
+	log.Info().Msg("Stop custom projection")
+	return err
+}
+
+func (p *CustomProjection) process(ctx context.Context, stream *esdb.PersistentSubscription) error {
+	log := zlog.Ctx(ctx)
+
+	for {
+		esdbEvent := stream.Recv()
+
+		select {
+		case <-ctx.Done():
+			// Return when context gets finished
+			return nil
+		default:
+		}
+
+		if esdbEvent.SubscriptionDropped != nil {
+			log.Error().Err(esdbEvent.SubscriptionDropped.Error).Msg("Subscription dropped")
+			return esdbEvent.SubscriptionDropped.Error
+		}
+
+		if esdbEvent.EventAppeared == nil || esdbEvent.EventAppeared.Event == nil {
+			log.Debug().Msg("Received empty event, skip")
+			continue
+		}
+
+		if strings.HasPrefix(esdbEvent.EventAppeared.Event.EventType, "$") {
+			// Skip internal events
+			// https://developers.eventstore.com/server/v23.10/streams.html#reserved-names
+			continue
+		}
+
+		event, err := hwes.NewEventFromRecordedEvent(esdbEvent.EventAppeared.Event)
+
+		if err != nil {
+			log.Error().
+				Err(err).
+				Str("EventType", esdbEvent.EventAppeared.Event.EventType).
+				Msg("could not create new event from recorded event")
+			continue
+		}
+
+		event.Log(log.Debug()).Msg("process event")
+
+		if err, nackAction := p.handleEvent(ctx, event); err != nil {
+			event.Log(log.Error()).Err(err).Msg("error during processing of event")
+
+			event.Log(log.Warn()).Msg("nack event")
```

Comment:
using zerlog.Dict() (or zlog.Dict(), not sure) might be more straight forward

Reply:
Very nice! Didn't know `zerolog.Dict()` 👍